share-coordinator/src/main/java/org/apache/kafka/coordinator/share/metrics/ShareCoordinatorMetrics.java

```diff
@@ -153,4 +158,31 @@ public void record(String sensorName) {
             globalSensors.get(sensorName).record();
         }
     }
+
+    public void recordPrune(double value, TopicPartition tp) {
+        pruneMetrics.computeIfAbsent(tp, k -> new ShareGroupPruneMetrics(tp))
+            .pruneSensor.record(value);
+    }
+
+    private class ShareGroupPruneMetrics {
+        private final Sensor pruneSensor;
+
+        ShareGroupPruneMetrics(TopicPartition tp) {
+            String sensorNameSuffix = tp.toString();
+            Map<String, String> tags = Map.of(
+                "topic", tp.topic(),
+                "partition", Integer.toString(tp.partition())
+            );
+
+            pruneSensor = metrics.sensor(SHARE_COORDINATOR_STATE_TOPIC_PRUNE_SENSOR_NAME + sensorNameSuffix);
+
+            pruneSensor.add(
+                metrics.metricName("last-pruned-offset",
```

Comment:
So essentially the metric is capturing the lowWaterMark for topic partition, hence what's the relevance of calling it `pruned`? Also what monitoring importance this metric captures?

Reply:
So how this metric differs from existing: `kafka.log:type=Log,name=LogStartOffset,topic=__share_group_state,partition=<partition_num>`

<img width="895" alt="Screenshot 2024-12-17 at 00 11 17" src="https://github.com/user-attachments/assets/28a90c78-cb39-48f1-ab73-df34a9ff5205" />
<img width="1216" alt="Screenshot 2024-12-17 at 00 12 21" src="https://github.com/user-attachments/assets/f8e8885d-09e8-4a17-b881-0811016d5475" />

Reply:
Like the case I mentioned our last pruned offset could be different we what the low water mark is (if we send something smaller). Current `Partition.scala` code simply swallows this condition with no specific indication.

This metric provides pruners perspective, low water mark/LogStartOffset provide replicamanager/partition/Unified log perspectives.